File: PA2/Lex.c
// Name: Bobby Singh Dhillon
// CruzID: bosdhill
// ID number: 1594478
// Assignment Name: Programming Assignment 2 -- List ADT
// Description: 
//-----------------------------------------------------------------------------
// Lex.c
// Client for List ADT
//-----------------------------------------------------------------------------

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "List.h"

// define aribitrary max length per string 
#define MAX_LEN 160

int main(int argc, char* argv[]){

	// File pointers 
	FILE *inFile, *outFile; 
	char strIn[MAX_LEN];
	int arrLen = 0;  

	// if command line doesn't have 3 arguments
	if (argc != 3){
		printf("Usage: %s <input file> <output file>\n", argv[0]);
		exit(1);
	}
	// open input file and output file 
	inFile = fopen(argv[1], "r"); 
	outFile = fopen(argv[2], "w");

	// check if inFile and outFile are pointing to NULL
	if(inFile == NULL){
		printf("Unable to open file %s for reading\n", argv[1]);
		exit(1);
	}
	if(outFile == NULL){
		printf("Unable to open file %s for writing\n", argv[2]);
		exit(1);
	} 
    // while its not the end of file
	while(!feof(inFile)){
   	// read into string 
		fgets(strIn, MAX_LEN, inFile);
 		arrLen++;  // inc count for size of array 
   	}

   	// create an array of char pointers 
   	char strArr[arrLen][MAX_LEN]; 

   // reset file pointer to beginning 
   	fseek(inFile, 0, SEEK_SET); 


   	// read in strings into the character Array 
   	for (int i = 0; i < arrLen - 1; i++) {
   		// read in strings to string Array
   		fgets(strIn, MAX_LEN, inFile); 
   		strcpy(strArr[i], strIn); 
   	}

   	// create a new List
   	List L = newList(); 

   	// start at index 0
   	append(L, 0); 
   	int j = 0; // for index in L
   	int k = 0; // result of strcmp

   	// sort indirectly 
   	for (int i = 1; i < arrLen; i++){
   		// move cursor to back
   		moveBack(L); 
   		while (index(L) >= 0){
   			// get index of L
   			j = get(L); 
			k = strcmp(strArr[i],strArr[j]); 

			// if current string is less than or 
			// equal to string of index List
   			if (k < 0){
   				// move down the list 
   				movePrev(L); 
   			}
   			else
   				break; 
   		}

   		// if cursor is undefined, prepend
   		if (index(L) == -1) 
   			prepend(L, i); 
   		// else if index is still zero and string
   		// is less than string in L, prepend
   		else if (index(L) == 0 && k < 0)
   			prepend(L, i); 
   		else // otherwise, insert after  
   			insertAfter(L, i); 
	}

	// move to front of List
	moveFront(L); 

	// write sorted array to outFile
	for (int i = 0; i < length(L); i++){
   		fprintf(outFile,"%s", strArr[get(L)]);
   		moveNext(L); 
	}

   	fclose(inFile); 
   	fclose(outFile); 

   	freeList(&L);

   return(0);
}

File: PA1/Lex.java
// Assignment Name: Programming Assignment 1 -- List ADT
// File: Lex.java 
// Description: 
//-----------------------------------------------------------------------------
// Lex.java is the client that implements the List ADT  in order to sort the 
// elements of an input array, read as strings from a file, indirectly. 
//-----------------------------------------------------------------------------


import java.io.*; 
import java.util.*; 

public class Lex{

	public static void main(String[] args) throws IOException{

		// first check if args.length is within bounds
		if (args.length != 2) {
         // report to std err
			System.err.println("Warning, insufficient command " + 
				"line arguments");
			System.exit(420);  
		}

        // intialize inFile and outFile objects with args[0] and args[1]
        // being the file names

		Scanner inFile = new Scanner(new File(args[0]));
		PrintWriter outFile = new PrintWriter(new FileWriter(args[1]));  
		String [] strArr = null; 


		// create a new array list 
		ArrayList<String> readIn = new ArrayList<String>(); 

		while (inFile.hasNextLine()) {

			// read in line and add to array list
			readIn.add(inFile.nextLine());
		}

		// get array version of all elements 
		strArr = new String[readIn.size()]; 
		strArr = readIn.toArray(strArr);

		// create a new List
		List L = new List(); 

		// start at index 0
		L.append(0); 
		int j = 0; 

		for (int i = 1; i < strArr.length; i++) {

			// move cursor to back
			L.moveBack(); 

			while (L.index() >= 0) {

				// get current index in List
				j = L.get();

				// if current string is less than or equal to List string
				if (strArr[i].compareTo(strArr[j]) < 0){

					// move down the list 
					L.movePrev(); 
				}
				else
					break; 
			}

			// if cursor is undefined, that means it was less than everything
			// in List, so it will be the new front 
			if (L.index() == -1){
				L.prepend(i); 
			} // else if index is still 0 and the str[i] is less than front
			// it becomes new front 
			else if (L.index() == 0 && strArr[i].compareTo(strArr[j]) < 0)
				L.prepend(i); 
			else // otherwise, just insert it after whatever it was not less than
				L.insertAfter(i); 

		}


	  	// start at front 
		L.moveFront(); 

      	// writing to file
		for (int i = 0; i < L.length(); i++) {
         	// write to file 
			outFile.println(strArr[L.get()]); 
			L.moveNext(); 
		}

		inFile.close();
		outFile.close();
	}
}
